monitoring/v1/types.go

```diff
@@ -0,0 +1,357 @@
+// +k8s:openapi-gen=true
+type AlertingRuleSpec struct {
+	// groups is a list of grouped alerting rules.  Rule groups are the unit at
+	// which Prometheus parallelizes rule processing.  All rules in a single group
+	// share a configured evaluation interval.  All rules in the group will be
+	// processed together on this interval, sequentially, and all rules will be
+	// processed.
+	//
+	// It's common to group related alerting rules into a single AlertingRule
+	// resources, and within that resource, closely related alerts, or simply
+	// alerts with the same interval, into individual groups.  You are also free
+	// to create AlertingRule resources with only a single rule group, but be
+	// aware that this can have a performance impact on Prometheus if the group is
+	// extremely large or has very complex query expressions to evaluate.
+	// Spreading very complex rules across multiple groups to allow them to be
+	// processed in parallel is also a common use-case.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MinItems:=1
+	// +kubebuilder:validation:Required
+	Groups []RuleGroup `json:"groups"`
+}
+
+// Duration is a valid prometheus time duration.
+// Supported units: y, w, d, h, m, s, ms
+// Examples: `30s`, `1m`, `1h20m15s`, `15d`
+// +kubebuilder:validation:Pattern:="^(0|(([0-9]+)y)?(([0-9]+)w)?(([0-9]+)d)?(([0-9]+)h)?(([0-9]+)m)?(([0-9]+)s)?(([0-9]+)ms)?)$"
+type Duration string
+
+// RuleGroup is a list of sequentially evaluated alerting rules.
+//
+// +k8s:openapi-gen=true
+type RuleGroup struct {
+	// name is the name of the group.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name"`
+
+	// interval is how often rules in the group are evaluated.  If not specified,
+	// it defaults to the global.evaluation_interval configured in Prometheus,
+	// which itself defaults to 30 seconds.  You can check if this value has been
+	// modified from the default on your cluster by inspecting the platform
+	// Prometheus configuration:
+	//
+	// $ oc -n openshift-monitoring describe prometheus k8s
+	//
+	// The relevant field in that resource is: spec.evaluationInterval
+	//
+	// +optional
+	Interval Duration `json:"interval,omitempty"`
+
+	// rules is a list of sequentially evaluated alerting rules.  Prometheus may
+	// process rule groups in parallel, but rules within a single group are always
+	// processed sequentially, and all rules are processed.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	// +kubebuilder:validation:Required
+	Rules []Rule `json:"rules"`
+}
+
+// Rule describes an alerting rule.
+// See Prometheus documentation:
+// - https://www.prometheus.io/docs/prometheus/latest/configuration/alerting_rules
+//
+// +k8s:openapi-gen=true
+type Rule struct {
+	// alert is the name of the alert. Must be a valid label value, i.e. may
+	// contain any Unicode character.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
```

Comment:
Should we set a max length? Even if it is pretty large (1024?), just to be safe

Reply:
I'm ok with that, having unbounded strings though could be bad, we should have some sensible but high limit on most things that have length

Reply:
:+1: for string types.
For lists and maps I think we need to be more careful with this limit. A string length limitation can possibly be worked around, not so for lists and maps.

Reply:
May contain any unicode character? Really? Are we sure we want to support that? Is there any way could validate this with a regex?

Reply:
I'm not sure why we want to put restrictions on which characters are supported knowing that Prometheus supports any valid Go string.

Reply:
In that case, fine, just wanted to ask the question a few times in case it triggered some realisation that actually, there are some characters we don't want to support, if you are certain any valid go string is good, then this is ok to proceed as is